text/0496-handlebars-strict-mode.md

```diff
@@ -0,0 +1,591 @@
+RFC, because it makes it impossible to pass helpers around as values, just as
+it is possible to pass around contextual components today.
+
+The contextaul helper RFC proposed to deprecate this behavior and require
+mandatory parentheses to invoke the helper.
+
+
+```hbs
+{{!-- passing the helper pi to the component --}}
+<MyComponent @value={{pi}} />
+
+{{!-- passing the result of invoking the helper pi to the component --}}
+<MyComponent @value={{(pi)}} />
+```
+
+In strict mode, the current, soon-to-be-deprecated behavior will be removed,
+and the parentheses syntax will be mandatory.
+
+Note that this only affects arguments-less helpers, which are exceedingly rare,
+as most helpers perform self-contained computations based on the provided
+arguments. It also only affect argument positions. In content and attribute
+positions, the intent is clear as it does not make sense to "pass a helper into
+the DOM". The parentheses-less form will continue to work in those positions,
+although the explicit parentheses are also permitted.
+
+### 4. No dynamic resolution
+
+Today, Ember supports passing strings to the `component` helper (as well as the
+`helper` and `modifier` helpers proposed in [RFC #432](https://github.com/emberjs/rfcs/pull/432)).
+This can either be passed as a literal `{{component "foo-bar"}}` or passed as a
+runtime value `{{component this.someString}}`. In either case, Ember will
+attempt to _resolve_ the string as a component.
+
+It shares some of the same problems with implicit globals (where did this come
+from?), but the dynamic form makes the problem more acute, as it is difficult
+or impossible to tell which components a given template is dependent on. As
+usual, if it is difficult for the human reader, the same is true for tools as
+well. Specifically, this is hostile to "tree shaking" and other forms of static
+(build time) dependency-graph analysis, since the dynamic form of the component
+helper can invoke _any_ component available to the app.
+
+We propose to remove support for these forms of dynamic resolutions in strict
+mode. Specifically, passing a string to the `component` helper (as well as the
+`helper` and `modifier` helpers), whether as a literal or at runtime, will
+result in an error.
+
+In practice, it is almost always the case that these dynamic resolutions are
+switching between a small and bounded number of known components. For this
+purpose, they can be replaced by patterns similar to this.
+
+```js
+import Component from '@glimmer/component';
+import { First, Second, Third } from './contextual-components';
+
+class ProviderComponent extends Component {
+  get selectedComponent() {
+    switch(this.args.selection) {
+      case "first":
+        return First;
+      case "second":
+        return Second;
+      case "third":
+        return Third;
+    }
+  }
+}
+```
+
+```hbs
+{{yield this.selectedComponent}}
+```
+
+This will make it clear to the human reader and enable tools to perform
```

Comment:
How will this work with template-only components? We have things like this quite often in our app, in template only components (simplified):

```hbs
{{yield
  (hash Item=(component 'my-item') Header=(component 'my-header'))
}}
```

It would say it would be less than ideal to require us to provide a backing class for all of these components just to pass down all the yielded sub components, especially as we've just been working on making as many components template-only as possible.

Reply:
Or more specifically, I guess this will be solved with template imports at some point (probably) - but they aren't really mentioned in this RFC as far as I can tell, so I was wondering if 

a) this would only ship after template imports
b) this would ship (potentially) before, but if you'd need this behavior you'd either need to wait on template imports to ship or create a backing JS class everywhere in the meanwhile
c) there would be an alternative way to achieve this - not that uncommon, I suspect - use case

I would of definitely prefer c) ;)

Reply:
Templates are compiled down to JavaScript in the end, so there already exist some artifacts that make them importable. Since the [Component Template Co-Location RFC](https://github.com/emberjs/rfcs/blob/master/text/0481-component-templates-co-location.md#template-only-components), this structure has been defined more thoroughly. Any template import preprocessor will be able to output something along the lines of:

```js
export default setComponentTemplate(
  precompileTemplate('Hello, {{name}}!'),
  templateOnlyComponent()
);
```

And that would define a TO component which can be imported like any other component which has a backing class. The co-location RFC effectively already does this for existing TO components.

How template import preprocessors do this, exactly, is yet to be defined. The goal of this RFC is to provide a public primitive that template import preprocessors and syntax can be built on top of, not the actual import syntax itself. That will come in a future RFC.

Reply:
I think @mydea is asking how to do the contextual components pattern without introducing a JavaScript class. Using this hypothetical template import syntax:

```hbs
---
import MyItem form '...';
import MyHeader from '...';
---

{{yield (hash Item=MyItem Header=MyHeader}}
```

For the example given in the RFC:

```hbs
---
import First from '...';
import Second from '...';
import Third from '...';
---

{{#if (eq @selected "first")}}
  {{yield First}}
{{else if (eq @selected "second")}}
  {{yield Second}}
{{else if (eq @selected "third")}}
  {{yield Third}}
{{/if}}
```

...or...

```hbs
---
import First from '...';
import Second from '...';
import Third from '...';
---

{{yield (get (hash first=First second=Second third=Third) @selected)}}
```

And for your second question, yes, this will ship first, but it is not really usable for end-users until template-import (or SFC) ships, which will be the primary intended way to "consume" strict mode. There are bigger problems than this pattern – you can't even invoke a regular component without importing it, so there isn't really a good way to use it without some surface syntax that works with import semantics. The plan is to land the primitive and being experimenting with these ideas, and eventually adopt one or more of them into mainline Ember.

Reply:
Quick clarification: would you be able to pass the imported component to the `component` helper in order to curry arguments to the component yielded?

E.g.

```
---
import Item from '...';
---

{{yield (hash item=(component Item arg=something))}}
```

Reply:
I believe the answer is yes here, because I'm pretty sure this is already something that works with the `{{component}}` helper (would need to double check to be sure though)

Reply:
Yep. It doesn't work today yet, but it was added in the contextual helper + co-location RFCs. Or rather, the component helper has and will always be happy to take an invokable component as the first argument; the contextual helper + co-location RFCs stated the goal to make the component JS class itself invokable, which follows that the above will work.